src/Channels.Networking.Sockets/SocketConnection.cs

```diff
@@ -242,104 +263,110 @@ private enum BufferStyle
-                                // We need  to do a speculative receive with a *cheap* buffer while we wait for input; it would be *nice* if
-                                // we could do a zero-length receive, but this is not supported equally on all platforms (fine on Windows, but
-                                // linux hates it). The key aim here is to make sure that we don't tie up an entire block from the memory pool
-                                // waiting for input on a socket; fine for 1 socket, not so fine for 100,000 sockets
-
-                                // do a short receive while we wait (async) for data
-                                initialDataBuffer = LeaseSmallBuffer();
-                                args.SetBuffer(initialDataBuffer, 0, initialDataBuffer.Length);
-                                if (Socket.ReceiveAsync(args))
-                                {
-                                    // wait async for the io work to be completed
-                                    await ((Signal)args.UserToken).WaitAsync();
-                                }
-                                break;
-                        }
-                        if (args.SocketError != SocketError.Success)
-                        {
-                            throw new SocketException((int)args.SocketError);
-                        }
 
-                        // note we can't check BytesTransferred <= 0, as we always
-                        // expect 0; but if we returned, we expect data to be
-                        // buffered *on the socket*, else EOF
-                        if ((bytesFromInitialDataBuffer = args.BytesTransferred) <= 0)
+                        int bytesFromInitialDataBuffer = 0;
+
+                        if (Socket.Available == 0)
                         {
-                            if ((object)initialDataBuffer == (object)_zeroLengthBuffer)
+                            // now, this gets a bit messy unfortunately, because support for the ideal option
+                            // (zero-length reads) is platform dependent
+                            switch (_bufferStyle)
                             {
-                                // sentinel value that means we should just
-                                // consume sync (we expect there to be data)
-                                initialDataBuffer = null;
+                                case BufferStyle.Unknown:
+                                    try
+                                    {
+                                        initialSegment = await ReceiveInitialDataUnknownStrategyAsync(args);
+                                    }
+                                    catch
+                                    {
+                                        initialSegment = default(ArraySegment<byte>);
+                                    }
+                                    if (initialSegment.Array == null)
```

Comment:
Yes; if the API blows up in a way that tells us we need to switch to small-buffers, `ReceiveInitialDataUnknownStrategyAsync`  returns `default(ArraySegment<byte>)`